fairmq/shmem/Message.h

```diff
@@ -291,21 +292,27 @@ class Message final : public fair::mq::Message
             fManager.GetSegment(fMeta.fSegmentId);
             ShmHeader::IncrementRefCount(fManager.GetAddressFromHandle(fMeta.fHandle, fMeta.fSegmentId));
         } else { // unmanaged region
-            if (otherMsg.fMeta.fShared < 0) { // if UR msg is not yet shared
-                // TODO: minimize the size to 0 and don't create extra space for user buffer alignment
-                char* ptr = fManager.Allocate(2, 0);
-                // point the fShared in the unmanaged region message to the refCount holder
-                otherMsg.fMeta.fShared = fManager.GetHandleFromAddress(ptr, fMeta.fSegmentId);
-                // the message needs to be able to locate in which segment the refCount is stored
-                otherMsg.fMeta.fSegmentId = fMeta.fSegmentId;
+            if (otherMsg.fMeta.fShared < 0) {
+                fRegionPtr = fManager.GetRegionFromCache(otherMsg.fMeta.fRegionId);
+                if (!fRegionPtr) {
+                    throw TransportError(tools::ToString("Cannot get unmanaged region with id ", fMeta.fRegionId));
+                }
+                // create a reference couting object in for this message
+                RefCount& rcPtr = fRegionPtr->MakeRefCount();
+                // point the fShared in the unmanaged region message to the refCount object
+                otherMsg.fMeta.fShared = fRegionPtr->HandleFromAddress(&rcPtr);
                 // point this message to the same content as the unmanaged region message
                 fMeta = otherMsg.fMeta;
                 // increment the refCount
-                ShmHeader::IncrementRefCount(ptr);
+                rcPtr.Increment();
             } else { // if the UR msg is already shared
                 fMeta = otherMsg.fMeta;
-                fManager.GetSegment(fMeta.fSegmentId);
-                ShmHeader::IncrementRefCount(fManager.GetAddressFromHandle(fMeta.fShared, fMeta.fSegmentId));
+                fRegionPtr = fManager.GetRegionFromCache(fMeta.fRegionId);
+                if (fRegionPtr) {
+                    fRegionPtr->GetRefCountAddressFromHandle(fMeta.fShared)->Increment();
+                } else {
+                    throw TransportError(tools::ToString("Cannot get unmanaged region with id ", fMeta.fRegionId));
+                }
             }
         }
     }
```

Comment:
One more optimization by initializing the ref count with `2` and dedup `Copy` a bit for readability:

```diff
--- a/fairmq/shmem/Message.h
+++ b/fairmq/shmem/Message.h
@@ -276,45 +276,37 @@ class Message final : public fair::mq::Message
     void Copy(const fair::mq::Message& other) override
     {
         const Message& otherMsg = static_cast<const Message&>(other);
+        // if the other message is not initialized, close this one too and return
         if (otherMsg.fMeta.fHandle < 0) {
-            // if the other message is not initialized, close this one too and return
             CloseMessage();
             return;
         }

+        // if this msg is already initialized, close it first
         if (fMeta.fHandle >= 0) {
-            // if this msg is already initialized, close it first
             CloseMessage();
         }

-        if (otherMsg.fMeta.fManaged) { // managed segment
-            fMeta = otherMsg.fMeta;
-            fManager.GetSegment(fMeta.fSegmentId);
-            ShmHeader::IncrementRefCount(fManager.GetAddressFromHandle(fMeta.fHandle, fMeta.fSegmentId));
-        } else { // unmanaged region
+        // increment ref count
+        if (otherMsg.fMeta.fManaged) { // msg in managed segment
+            fManager.GetSegment(otherMsg.fMeta.fSegmentId);
+            ShmHeader::IncrementRefCount(fManager.GetAddressFromHandle(otherMsg.fMeta.fHandle, otherMsg.fMeta.fSegmentId));
+        } else { // msg in unmanaged region
+            fRegionPtr = fManager.GetRegionFromCache(otherMsg.fMeta.fRegionId);
+            if (!fRegionPtr) {
+                throw TransportError(tools::ToString("Cannot get unmanaged region with id ", fMeta.fRegionId));
+            }
+
             if (otherMsg.fMeta.fShared < 0) {
-                fRegionPtr = fManager.GetRegionFromCache(otherMsg.fMeta.fRegionId);
-                if (!fRegionPtr) {
-                    throw TransportError(tools::ToString("Cannot get unmanaged region with id ", fMeta.fRegionId));
-                }
-                // create a reference couting object in for this message
-                RefCount& rcPtr = fRegionPtr->MakeRefCount();
-                // point the fShared in the unmanaged region message to the refCount object
-                otherMsg.fMeta.fShared = fRegionPtr->HandleFromAddress(&rcPtr);
-                // point this message to the same content as the unmanaged region message
-                fMeta = otherMsg.fMeta;
-                // increment the refCount
-                rcPtr.Increment();
-            } else { // if the UR msg is already shared
-                fMeta = otherMsg.fMeta;
-                fRegionPtr = fManager.GetRegionFromCache(fMeta.fRegionId);
-                if (fRegionPtr) {
-                    fRegionPtr->GetRefCountAddressFromHandle(fMeta.fShared)->Increment();
-                } else {
-                    throw TransportError(tools::ToString("Cannot get unmanaged region with id ", fMeta.fRegionId));
-                }
+                // UR msg not yet shared, create the reference counting object with count 2
+                otherMsg.fMeta.fShared = fRegionPtr->HandleFromAddress(&(fRegionPtr->MakeRefCount(2)));
+            } else {
+                fRegionPtr->GetRefCountAddressFromHandle(otherMsg.fMeta.fShared)->Increment();
             }
         }
+
+        // copy meta data
+        fMeta = otherMsg.fMeta;
     }

     ~Message() override { CloseMessage(); }
diff --git a/fairmq/shmem/UnmanagedRegion.h b/fairmq/shmem/UnmanagedRegion.h
index 539a93822..a1b147f5c 100644
--- a/fairmq/shmem/UnmanagedRegion.h
+++ b/fairmq/shmem/UnmanagedRegion.h
@@ -187,10 +187,10 @@ struct UnmanagedRegion

     bool RemoveOnDestruction() { return fRemoveOnDestruction; }

-    RefCount& MakeRefCount()
+    RefCount& MakeRefCount(uint16_t initial_count = 1)
     {
         RefCount* refCount = fRefCountPool->allocate(1).get();
-        new (refCount) RefCount(1);
+        new (refCount) RefCount(initial_count);
         return *refCount;
     }

```

Reply:
Nice, updated.